master/pkg/searcher/asha_stopping.go

```diff
@@ -112,111 +132,113 @@ func (s *asyncHalvingStoppingSearch) initialOperations(ctx context) ([]Operation
-	if rungIndex == s.NumRungs()-1 {
-		rung.Metrics = append(rung.Metrics,
-			trialMetric{
-				RequestID: requestID,
-				Metric:    model.ExtendedFloat64(metric),
-			},
-		)
+// stopRun handles early-stopping and record-keeping logic for a validation metric reported to the
+// searcher.
+// If the metric qualifies the run for a rung but is not in the top 1/divisor trials for that rung,
+// stopRun will return a single `searcher.Stop` action. Otherwise, no actions will be returned.
+func (s *asyncHalvingStoppingSearch) stopRun(
+	trialID int32, timeStep uint64, metric float64,
+) []Action {
+	rungIndex := s.TrialRungs[trialID]
+	var actions []Action
+
+	// Starting at current rung, check if run should continue to next rung or early-stop.
+	// Since validations aren't controlled by searcher, they could complete > 1 rungs at a time.
+	for r := rungIndex; r < s.NumRungs(); r++ {
+		rung := s.Rungs[r]
+		s.TrialRungs[trialID] = r
 
-		if !s.EarlyExitTrials[requestID] {
-			ops = append(ops, NewClose(requestID))
-			s.ClosedTrials[requestID] = true
+		// If run has not completed enough steps to qualify for this rung, exit.
+		if timeStep < rung.UnitsNeeded {
+			return actions
 		}
-	} else {
-		// This is not the top rung, so do promotions to the next rung.
-		nextRung := s.Rungs[rungIndex+1]
-		// We need to run continueTraining even if the trial was terminated early so that we
-		// can add the metric to the rung.
-		promoteTrial := rung.continueTraining(
-			requestID,
-			metric,
-			s.Divisor(),
-		)
-		// In contrast to promotion-based ASHA, we will not let early-exited trials add
-		// -/+inf metrics to higher rungs even if portion of terminated trials in bottom rung
-		// is greater than 1 - 1 / divisor.
-		if !s.EarlyExitTrials[requestID] {
-			if promoteTrial {
-				s.TrialRungs[requestID] = rungIndex + 1
-				nextRung.OutstandingTrials++
-				unitsNeeded := mathx.Max(nextRung.UnitsNeeded-rung.UnitsNeeded, 1)
-				ops = append(ops, NewValidateAfter(requestID, unitsNeeded))
-				addedTrainWorkload = true
-			} else {
-				ops = append(ops, NewClose(requestID))
-				s.ClosedTrials[requestID] = true
-			}
+
+		insertIndex := rung.insertMetric(trialID, metric)
+
+		// If this is the top rung, close the run and exit.
+		if r == s.NumRungs()-1 {
+			actions = append(actions, NewStop(trialID))
+			return actions
 		}
-	}
 
-	allTrials := len(s.TrialRungs) - s.InvalidTrials
-	if !addedTrainWorkload && allTrials < s.MaxTrials() {
-		create := NewCreate(
-			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
-		s.TrialRungs[create.RequestID] = 0
-		ops = append(ops, create)
-		ops = append(ops, NewValidateAfter(create.RequestID, s.Rungs[0].UnitsNeeded))
-	}
+		// Top 1/divisor trials should continue, trials - 1/divisor trials should be stopped.
+		// If trials < divisor, continue only if this is the best performing run so far.
+		numContinue := mathx.Max(int(float64(len(rung.Metrics))/s.Divisor()), 1)
```

Comment:
why is the original algorithm was
```suggestion
		numContinue := mathx.Max(int(float64(len(rung.Metrics) + 1)/s.Divisor()), 1)
```
and yours drops the + 1? i guess the max protects us from an aggressive divisor not promoting anything to the last rung. surely liam would know? some weird floating point protection, haha?

Reply:
i dropped it simply because before, this was calculated before the new metric was added, hence the +1 to the metric count.

Reply:
that may be all it is haha. i just was reading the new code as careful as i could for any weird differences.